google/resource_dataflow_job.go

```diff
@@ -0,0 +1,210 @@
+}
+
+func resourceDataflowJob() *schema.Resource {
+	return &schema.Resource{
+		Create: resourceDataflowJobCreate,
+		Read:   resourceDataflowJobRead,
+		Delete: resourceDataflowJobDelete,
+
+		Schema: map[string]*schema.Schema{
+			"name": &schema.Schema{
+				Type:     schema.TypeString,
+				Required: true,
+				ForceNew: true,
+			},
+
+			"template_gcs_path": &schema.Schema{
+				Type:     schema.TypeString,
+				Required: true,
+				ForceNew: true,
+			},
+
+			"temp_gcs_location": &schema.Schema{
+				Type:     schema.TypeString,
+				Required: true,
+				ForceNew: true,
+			},
+
+			"zone": &schema.Schema{
+				Type:     schema.TypeString,
+				Optional: true,
+				ForceNew: true,
+			},
+
+			"max_workers": &schema.Schema{
+				Type:     schema.TypeInt,
+				Optional: true,
+				ForceNew: true,
+			},
+
+			"parameters": {
+				Type:     schema.TypeMap,
+				Optional: true,
+				ForceNew: true,
+			},
+
+			"on_delete": &schema.Schema{
+				Type:         schema.TypeString,
+				ValidateFunc: validation.StringInSlice([]string{"cancel", "drain"}, false),
+				Optional:     true,
+				Default:      "drain",
+				ForceNew:     true,
+			},
+
+			"project": &schema.Schema{
+				Type:     schema.TypeString,
+				Optional: true,
+				ForceNew: true,
+			},
+
+			"state": &schema.Schema{
+				Type:     schema.TypeString,
+				Computed: true,
+			},
+		},
+	}
+}
+
+func resourceDataflowJobCreate(d *schema.ResourceData, meta interface{}) error {
+	config := meta.(*Config)
+
+	project, err := getProject(d, config)
+	if err != nil {
+		return err
+	}
+
+	jobName := d.Get("name").(string)
+	gcsPath := d.Get("template_gcs_path").(string)
+	tempLocation := d.Get("temp_gcs_location").(string)
+	zone := d.Get("zone").(string)
+	maxWorkers := d.Get("max_workers").(int)
+	params := expandDataflowParamsStringMap(d.Get("parameters").(map[string]interface{}))
+
+	env := dataflow.RuntimeEnvironment{
+		TempLocation: tempLocation,
+		Zone:         zone,
+		MaxWorkers:   int64(maxWorkers),
+	}
+
+	request := dataflow.CreateJobFromTemplateRequest{
+		JobName:     jobName,
+		GcsPath:     gcsPath,
+		Parameters:  params,
+		Environment: &env,
+	}
+
+	job, err := config.clientDataflow.Projects.Templates.Create(project, &request).Do()
+	if err != nil {
+		return err
+	}
+
+	d.SetId(job.Id)
+
+	return resourceDataflowJobRead(d, meta)
+}
+
+func resourceDataflowJobRead(d *schema.ResourceData, meta interface{}) error {
+	config := meta.(*Config)
+
+	project, err := getProject(d, config)
+	if err != nil {
+		return err
+	}
+
+	id := d.Id()
+
+	job, err := config.clientDataflow.Projects.Jobs.Get(project, id).Do()
+	if err != nil {
+		return handleNotFoundError(err, d, fmt.Sprintf("Dataflow job %s", id))
+	}
+
+	if dataflowTerminalStatesMap[job.CurrentState] {
+		log.Printf("[DEBUG] Removing resource '%s' because it is in state %s.\n", job.Name, job.CurrentState)
+		d.SetId("")
+		return nil
+	} else {
+		d.Set("state", job.CurrentState)
+	}
+
```

Comment:
Generally it's a good idea to set as many fields as you have access to on read, not just computed ones. This helps detect any drift between what Terraform thinks is true and what GCP thinks is true.

Reply:
Done.